File: 3-unconstrained-optimization/python/ex_quasi_newton.py
import numpy as np
import numpy.random as random
from numpy.linalg import norm,solve


# function to make it easier to display things in the command window
def disp_helper(name,number):
    
    # form string
    string = name + ' = {}'  
    
    # display string
    print(string.format(number))



test = 3
updateformulaflag = 1

if test == 1:
    Q = np.diag([2,3,4]) # x'*Q*x/2
    c = np.array([-8,-9,-8]) # c'x
    B = np.eye(len(Q)) # initial approximate hessian matrix
    x = np.zeros((len(Q),)) # initial point
    
elif test == 2:
    random.seed(3253) # set random seed
    q = random.random((3,3))
    Q = q + q.T # x'*Q*x/2
    c = np.array([-8,-9,-8]) # c'x
    B = np.eye(len(Q)) # initial approximate hessian matrix
    x = np.zeros((len(Q),)) # initial point
    
elif test == 3:
    random.seed(786796)
    n = 20
    q = random.random((n,n))
    Q = q + q.T + 8*np.eye(n) # x'*Q*x/2
    c = random.random(n) # c'x
    B = np.eye(n) # initial approximate hessian matrix
    x = np.zeros((n,)) # initial point
    
    
    
# exact gradient
G = lambda x: Q.dot(x) - c

# exact hessian
H = lambda x: Q

# maximum number of iterations
max_iterations = 100

# compute initial gradient
g = G(x)

# calculate eps
eps = np.finfo(float).eps

# go through each iteration
for i in range(max_iterations):
    
    # (i) check if xk is optional
    if norm(g,ord = np.inf) <= 100*eps:
        print('Optimal!')
        break
    else:
        disp_helper('iteration',i+1)
        disp_helper('norm(g)',norm(g,ord = np.inf))
        
    # (ii) solve for search direction
    p = -solve(B,g)
    disp_helper('p',p)

    # (iii) using exact line search 
    alpha = -(p.dot(g))/(p.dot(Q.dot(p)))
    disp_helper('alpha',alpha)
    
    x_new = x + alpha*p 
    disp_helper('x',x_new)
    
    # (iv) compute intermediate quantities
    g_new = G(x_new) # gradient
    s = x_new - x # step 
    y = g_new - g # gradient difference
    
    disp_helper('g',g_new)
    disp_helper('s',s)
    disp_helper('y',y)
    
    # (v) use symmetric rank-one update formula 
    if updateformulaflag == 1: #use symmetric rank-one update formula
        z = y - B.dot(s)
        z_ = z.reshape([-1,1])
        B_new = B + z_.dot(z_.T)/(z.dot(s))

    elif updateformulaflag == 2: # BFGS update formula
        z = B.dot(s)
        z_ = z.reshape([-1,1])
        y_ = y.reshape([-1,1])
        B_new = B - (z_.dot(z_.T))/(z.dot(s)) + (y_.dot(y_.T))/(y.dot(s))
        
    disp_helper('B',B_new)
    
    # update stored variables
    B = B_new
    g = g_new 
    x = x_new
    
    print('')
    

# compute error compared to exact solution
e_norm = norm(x - solve(Q,c),ord = np.inf)
disp_helper('max abs error',e_norm)

# compute error comprared to exact hessian
e_Q = norm(B-Q,ord = np.inf)
disp_helper('error Q',e_Q)            
    
    

File: 3-unconstrained-optimization/matlab/ex_quasi_newton.m
% ex_quasi_newton.m
% examples of the update formulas used in quasi-Newton methods including
% symmetric rank-one update and BFGS
% [reference] Section 12.3 in LNO
% [course] Session 8 - Unconstrained Optimization (2)
close all; clear; clc

example = 4;

updateformulaflag = 2; % 1: symmetric rank-one update, 2:BFGS

switch example
    %----------------------------------------------------------------------
    case 1 % quadratic function Example 12.9 in LNO
        quadraticFlag = true; % is this a quadratic function?
        Q = diag([2,3,4]); % x'*Q*x/2
        c = [-8;-9;-8]; % c'*x
        x0 = zeros(size(Q,1),1); % initial point
        B = eye(length(Q)); % initial approximate hessian matrix
    %----------------------------------------------------------------------
    case 2 % randomized 3D quadratic function example
        quadraticFlag = true; % is this a quadratic function?
        rng(3253) % set random seed
        q = rand(3);
        Q = q+q'; % x'*Q*x/2
        c = [-8;-9;-8]; % c'*x
        x0 = zeros(size(Q,1),1); % initial point
        B = eye(length(Q)); % initial approximate hessian matrix
    %----------------------------------------------------------------------
    case 3 % randomized nD quadratic function example
        quadraticFlag = true; % is this a quadratic function?
        rng(5754) % set random seed
        n = 20;
        q = rand(n);
        Q = q + q'+ 8*eye(n); % x'*Q*x/2
        c = rand(n,1); % c'*x
        x0 = zeros(n,1); % initial point
        B = eye(n); % initial approximate hessian matrix
    %----------------------------------------------------------------------
    case 4 % nonlinear 2D example
        quadraticFlag = false; % is this a quadratic function?
        n = 2; x = sym('x',[n 1]); % create symbolic variables
        f = x(1)^4 + 2*x(1)^3 + 24*x(1)^2 + x(2)^4 + 12*x(2)^2; % objective function
        B = eye(n); % initial approximate hessian matrix
        x0 = [2; 1]; % initial point
end

% create problem functions
if quadraticFlag
    G = @(x) Q*x - c; % exact gradient
    H = @(x) Q; % exact hessian
else
    g = gradient(f,x); % exact gradient
    h = hessian(f,x); % exact hessian
    F = matlabFunction(f,'Vars',{x}); % objective function
    G = matlabFunction(g,'Vars',{x}); % gradient
    H = matlabFunction(h,'Vars',{x}); % hessian
end

% maximum number of iterations
max_iterations = 100;

% compute initial gradient
g = G(x0);

% assign initial value
x = x0;

% go through each iteration
for iteration = 0:max_iterations-1

    % (i) check if xk is optimal
    if norm(g,inf) <= 100*(eps)
        disp('Optimal!')
        break % stop iterating
    else
        disp_helper("iteration",iteration)
        disp_helper("norm(g)",norm(g,inf))
    end

    % (ii) solve for the search direction
    p = -B\g;
    disp_helper("p",p)

    % form string
    str = strcat("B =",mat2str(round(B,3)));

    % display string
    disp(str)

    % (iii) determine step using line search
    if quadraticFlag % quadratic objective function exact line search
        alpha = -(p'*g)/(p'*Q*p);
    else % general case
        alpha = fminbnd(@(alpha) F(x+alpha*p),0,100,...
            optimset('TolX',1e-10,'TolFun',1e-10));
    end
    disp_helper("alpha",alpha)

    x_new = x + alpha*p; % take the step
    disp_helper("x",x_new)

    % (iv) compute intermediate quantities
    g_new = G(x_new); % gradient
    s = x_new - x; % step
    y = g_new - g; % gradient difference

    disp_helper("g",g_new)
    disp_helper("s",s)
    disp_helper("y",y)

    % (v) use a quasi-Newton update formula
    switch updateformulaflag
        case 1 % use symmetric rank-one update formula
            z = y - B*s;
            B_new = B + z*z'/(z'*s);
        case 2 % BFGS update formula
            z = B*s;
            B_new = B - z*z'/(z'*s) + y*y'/(y'*s);
    end
    disp_helper("B",B_new)

    % update stored variables
    B = B_new;
    g = g_new;
    x = x_new;
    disp(" ")

end

% compute error compared to the exact hessian
e_Q = norm(B-H(x),inf);
disp_helper("error Q",e_Q)

% compute error compared to the "exact" solution
if quadraticFlag
    e_norm = norm(x - (Q\c),inf);
    disp_helper("max abs error",e_norm)
end

%--------------------------------------------------------------------------
% function to make it easier to display things in the command window
function disp_helper(name,number)

% form string
str = strcat(string(name)," = ",mat2str(round(number,16)));

% display string
disp(str)

end